ingest/cell_metadata.py

```diff
@@ -1,39 +1,94 @@
-        self.metadata_types = self.file.readline().rstrip('\n').split('\t')
-        self.annotation_type = ['group', 'numeric']
-        self.errors = defaultdict(list)
-        self.ontology = defaultdict(lambda: defaultdict(set))
-        self.type = defaultdict(list)
+"""Module for ingesting cell metadata files
+
+DESCRIPTION
+Module provides extract and transform functions for cell metadata files.
+Text, CSV, and TSV files are supported.
+
+PREREQUISITES
+Must have python 3.6 or higher.
+"""
+import copy
+import json
+import os
+from typing import Dict, Generator, List, Tuple, Union
+
+from ingest_files import IngestFiles
+
+
+class CellMetadata(IngestFiles):
+    ALLOWED_FILE_TYPES = ['text/csv',
+                          'text/plain', 'text/tab-separated-values']
+
+    def __init__(self, file_path, file_id: str = None, study_accession: str = None):
+
+        IngestFiles.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
+        self.headers = self.get_next_line(
+            increase_line_count=False)
+        self.metadata_types = self.get_next_line(
+            increase_line_count=False)
+        # unique values for group-based annotations
+        self.uniqueValues = []
+        self.cell_names = []
+        self.top_level_doc = self.create_documents(
+            file_path, file_id, study_accession)
+        self.data_subcollection = self.create_subdocuments()
+        self.errors = {}
+
+    def transform(self, row: List[str]) -> None:
+        """ Add data from cell metadata files into data model"""
+        for idx, column in enumerate(row):
+            if idx != 0:
+                # if annotation is numeric convert from string to float
+                if self.metadata_types[idx].lower() == 'numeric':
+                    column = round(float(column), 3)
+                elif self.metadata_types[idx].lower() == 'group':
+                    # Check for unique values
+                    if column not in self.uniqueValues:
+                        self.uniqueValues.append(column)
+                # Get annotation name from header
+                annotation = self.headers[idx]
+                self.data_subcollection[annotation]['values'].append(column)
+            else:
+                # If column isn't a annotation value, it's a cell name
+                self.cell_names.append(column)
+
+    def create_documents(self, file_path, file_id, study_accession):
+        """Creats top level documents for Cell Metadata data structure"""
+        documents = {}
+
+        # Each annotation value has a top level document
+        for value in self.headers[1:]:
+            # Copy document model so memory references are different
+            copy_of_doc_model = copy.copy({
+                'name': value,
+                'study_accession': study_accession,
+                'source_file_name': file_path.strip("."),
+                'source_file_type': 'metadata',
```

Comment:
source_file_type actually isn't in the [Firestore Data Structures](https://docs.google.com/document/d/1ArIKzmJFUh-mjojGXquCjnJVGycrHfdOrrlAO-TG8cc/edit#heading=h.q45cv6vfiqrg) schema for CellMetadata documents - please remove.